dpe/src/commands/derive_child.rs

```diff
@@ -15,36 +15,122 @@ pub struct DeriveChildCmd {
+        self.flags & Self::CHANGE_LOCALITY != 0
+    }
+
+    /// Check if it will be able to create a new default.
+    ///
+    /// # Arguments
+    ///
+    /// * `parent_idx` - Index of the soon-to-be parent.
+    /// * `default_context_idx` - Index of the current default context, if there is one.
+    fn safe_to_make_default(&self, parent_idx: usize, default_context_idx: Option<usize>) -> bool {
+        if let Some(default_idx) = default_context_idx {
+            if default_idx != parent_idx || self.retains_parent() {
+                return false;
+            }
+        }
+        true
     }
 }
 
 impl<C: Crypto> CommandExecution<C> for DeriveChildCmd {
-    fn execute(&self, _dpe: &mut DpeInstance<C>, _locality: u32) -> Result<Response, DpeErrorCode> {
-        Err(DpeErrorCode::InvalidCommand)
+    fn execute(&self, dpe: &mut DpeInstance<C>, locality: u32) -> Result<Response, DpeErrorCode> {
+        // Make sure the operation is supported.
+        if !dpe.support.internal_info && self.uses_internal_info_input()
+            || !dpe.support.internal_dice && self.uses_internal_dice_input()
+        {
+            return Err(DpeErrorCode::InvalidArgument);
+        }
+
+        let parent_idx = dpe
+            .get_active_context_pos(&self.handle, locality)
+            .ok_or(DpeErrorCode::InvalidHandle)?;
+        let child_idx = dpe
+            .get_next_inactive_context_pos()
+            .ok_or(DpeErrorCode::MaxTcis)?;
+
+        let target_locality = if !self.changes_locality() {
+            locality
+        } else {
+            // Make sure the target locality is valid.
+            if !dpe.localities.iter().any(|&l| l == self.target_locality) {
+                return Err(DpeErrorCode::InvalidLocality);
+            }
+            self.target_locality
+        };
+
+        // Make sure it can be the default if it is supposed to be.
+        if self.makes_default() {
+            let default_context_idx =
+                dpe.get_active_context_pos(&ContextHandle::default(), target_locality);
+
+            if !self.safe_to_make_default(parent_idx, default_context_idx) {
+                return Err(DpeErrorCode::InvalidArgument);
+            }
+        }
+
+        dpe.add_tci_measurement(child_idx, &TciMeasurement(self.data))?;
+
+        let child_handle = if self.makes_default() {
+            ContextHandle::default()
+        } else {
+            dpe.generate_new_handle()?
+        };
+
+        let parent = &mut dpe.contexts[parent_idx];
+
+        let reported_parent_handle = if !self.retains_parent() {
+            parent.state = ContextState::Retired;
+            // The value doesn't matter because the client should ignore it.
+            ContextHandle([0xff; ContextHandle::SIZE])
+        } else if parent.handle.is_default() {
+            parent.handle
+        } else {
+            dpe.contexts[parent_idx].handle = dpe.generate_new_handle()?;
+            dpe.contexts[parent_idx].handle
+        };
```

Comment:
Hm, if we can't use `parent` in the final else clause, may as well not use it in the other clauses, for consistency.

Now that I know more about borrow checking and scopes, what about this? (Just trying to avoid a tri-case 'if' statement)

```
if !self.retains_parent() {
  dpe.contexts[parent_idx].state = ContextState::Retired;
  dpe.contexts[parent_idx].handle = ContextHandle([0xff; ContextHandle::SIZE]);
} else if !dpe.contexts[parent_idx].handle.is_default() {
  dpe.contexts[parent_idx].handle = dpe.generate_new_handle()?;
}

let reported_parent_handle = dpe.contexts[parent_idx].handle;
// Or just use dpe.contexts[parent_idx].handle directly down below
```

Reply:
Ahh nice! You're right the middle condition is pretty useless.

Done